library/objective-c/EnvoyEngineImpl.m

```diff
@@ -10,6 +11,91 @@ static void ios_on_exit() {
   NSLog(@"[Envoy] library is exiting");
 }
 
+typedef struct {
+  __unsafe_unretained EnvoyHTTPFilter *filter;
+} ios_http_filter_context;
+
+// TODO(goaway): refactor and relocate translation
+static envoy_data toManagedNativeString(NSString *s) {
+  size_t length = s.length;
+  uint8_t *native_string = (uint8_t *)safe_malloc(sizeof(uint8_t) * length);
+  memcpy(native_string, s.UTF8String, length);
+  return (envoy_data){length, native_string, free, native_string};
+}
+
+static EnvoyHeaders *to_ios_headers(envoy_headers headers) {
+  NSMutableDictionary *headerDict = [NSMutableDictionary new];
+  for (envoy_header_size_t i = 0; i < headers.length; i++) {
+    envoy_header header = headers.headers[i];
+    NSString *headerKey = [[NSString alloc] initWithBytes:header.key.bytes
+                                                   length:header.key.length
+                                                 encoding:NSUTF8StringEncoding];
+    NSString *headerValue = [[NSString alloc] initWithBytes:header.value.bytes
+                                                     length:header.value.length
+                                                   encoding:NSUTF8StringEncoding];
+    NSMutableArray *headerValueList = headerDict[headerKey];
+    if (headerValueList == nil) {
+      headerValueList = [NSMutableArray new];
+      headerDict[headerKey] = headerValueList;
+    }
+    [headerValueList addObject:headerValue];
+  }
+  // TODO(goaway): consider solution that doesn't violate release convention
```

Comment:
I feel like the user can determine if they are going to modify or not, and this function should release on copy? Or conversely, change the existing API and have the caller bear the responsibility of releasing? In any case, I think we can unify these helpers with the ones in EnvoyHTTPStreamImpl

Reply:
The convention I feel we should maintain is that the callee always releases or passes that responsibility. I don't think it's really possible to invert that, because there's no way for the caller (or returner) to otherwise determine what the callee does with it. This is true for calls going in either direction across the bridge layer.

While I do think we can reconcile the differences here, I added it as a TODO because it felt a little bit extraneous to getting the core functionality in place.